src/autolabel/data_loaders/__init__.py

```diff
@@ -0,0 +1,155 @@
+import logging
+import pandas as pd
+from typing import Dict, Union
+from tabulate import tabulate
+from datasets import Dataset
+from autolabel.data_loaders.read_datasets import (
+    DataAttribute,
+    CSVReader,
+    JsonlReader,
+    HuggingFaceDataset,
+    # SqlDataset,
+    DataFrameDataset,
+)
+from autolabel.data_loaders.validation import TaskDataValidation
+from autolabel.configs import AutolabelConfig
+
+logger = logging.getLogger(__name__)
+
+
+class DataValidationFailed(Exception):
+    def __init__(self, message):
+        self.message = message
+        super().__init__(self.message)
+
+
+class DatasetLoader:
+    # TODO: add support for reading from SQL databases
+    # TODO: add support for reading and loading datasets in chunks
+    MAX_ERROR_DISPLAYED = 100
+
+    def __init__(
+        self,
+        dataset: Union[str, pd.DataFrame],
+        config: AutolabelConfig,
+        max_items: int = 0,
+        start_index: int = 0,
+    ) -> None:
+        """DatasetLoader class to read and load datasets.
+
+        Args:
+            dataset (Union[str, pd.DataFrame]): path to the dataset or the dataframe
+            config (AutolabelConfig): config object
+            max_items (int, optional): max number of items to read. Defaults to 0.
+            start_index (int, optional): start index to read from. Defaults to 0.
+        """
+        self.dataset = dataset
+        self.config = config
+        self.max_items = max_items
+        self.start_index = start_index
+
+        self.__data_attr: DataAttribute = None
+        self.__malformed_records = None
+        self._read()
+
+    @property
+    def dat(
+        self,
+    ) -> Union[pd.DataFrame, Dataset]:
+        return self.__data_attr.dataset
+
+    @property
+    def inputs(
+        self,
+    ) -> Dict:
+        return self.__data_attr.inputs
+
+    @property
+    def gt_labels(
+        self,
+    ) -> str:
+        return self.__data_attr.gt_labels
+
+    @property
+    def columns(
+        self,
+    ) -> str:
+        return self.__data_attr.columns
+
+    def _read(
+        self,
+    ):
+        if isinstance(self.dataset, str):
+            self.__data_attr: DataAttribute = self._read_file(
+                self.dataset, self.config, self.max_items, self.start_index
+            )
+        elif isinstance(self.dataset, Dataset):
+            self.__data_attr: DataAttribute = HuggingFaceDataset.read(
+                self.dataset, self.config, self.max_items, self.start_index
+            )
+        elif isinstance(self.dataset, pd.DataFrame):
+            self.__data_attr: DataAttribute = DataFrameDataset.read(
+                self.dataset, self.config, self.start_index, self.max_items
+            )
+
+    def _read_file(
```

Comment:
we can merge this with the if/else in the caller function - don't need a new function for reading file